scio-core/src/main/scala/com/spotify/scio/sql/Sql.scala

```diff
@@ -0,0 +1,341 @@
+                                 bTag: TupleTag[B],
+                                 udfs: List[Udf] = Nil)
+
+object Queries {
+
+  /**
+    * Typecheck [[Query]] q against the provided schemas.
+    * If the query correctly typechecks, it's simply return as a [[Right]].
+    * If it fails, a error message is returned in a [[Left]].
+    */
+  def typecheck[A: Schema, B: Schema](q: Query[A, B]): Either[String, Query[A, B]] = {
+    val schema: BSchema = SchemaMaterializer.fieldType(Schema[A]).getRowSchema
+    val expectedSchema: BSchema =
+      Schema[B] match {
+        case s: Record[B] =>
+          SchemaMaterializer.fieldType(s).getRowSchema
+        case _ =>
+          SchemaMaterializer.fieldType(Schema[ScalarWrapper[B]]).getRowSchema
+      }
+
+    QueryUtils.typecheck(q.query, schema, expectedSchema).right.map(_ => q)
+  }
+
+  /**
+    * Typecheck [[Query2]] q against the provided schemas.
+    * If the query correctly typechecks, it's simply return as a [[Right]].
+    * If it fails, a error message is returned in a [[Left]].
+    */
+  def typecheck[A: Schema, B: Schema, R: Schema](
+    q: Query2[A, B, R]): Either[String, Query2[A, B, R]] = {
+    val schemaA: BSchema = SchemaMaterializer.fieldType(Schema[A]).getRowSchema
+    val schemaB: BSchema = SchemaMaterializer.fieldType(Schema[B]).getRowSchema
+    val expectedSchema: BSchema =
+      Schema[R] match {
+        case s: Record[R] =>
+          SchemaMaterializer.fieldType(s).getRowSchema
+        case _ =>
+          SchemaMaterializer.fieldType(Schema[ScalarWrapper[R]]).getRowSchema
+      }
+
+    QueryUtils
+      .typecheck(q.query, List((q.aTag.getId, schemaA), (q.bTag.getId, schemaB)), expectedSchema)
+      .right
+      .map(_ => q)
+  }
+
+  def typedQuery[A: Schema, B: Schema](query: String, udfs: Udf*): Query[A, B] =
+    macro QueryMacros.typedQuerylImpl[A, B]
+
+  def typedQuery[A: Schema, B: Schema, R: Schema](query: String,
+                                                  aTag: TupleTag[A],
+                                                  bTag: TupleTag[B],
+                                                  udfs: Udf*): Query2[A, B, R] =
+    macro QueryMacros.typedQuery2Impl[A, B, R]
+}
+
+object Sql {
+
+  def from[A: Schema](sc: SCollection[A]): SqlSCollection[A] = new SqlSCollection(sc)
+
+  def from[A: Schema, B: Schema](a: SCollection[A], b: SCollection[B]): SqlSCollection2[A, B] =
+    new SqlSCollection2(a, b)
+
+  private[sql] def registerUdf(t: SqlTransform, udfs: Udf*): SqlTransform =
+    udfs.foldLeft(t) {
+      case (st, x: UdfFromClass[_]) =>
+        st.registerUdf(x.fnName, x.clazz)
+      case (st, x: UdfFromSerializableFn[_, _]) =>
+        st.registerUdf(x.fnName, x.fn)
+      case (st, x: UdafFromCombineFn[_, _, _]) =>
+        st.registerUdaf(x.fnName, x.fn)
+    }
+
+}
+
+final class SqlSCollection[A: Schema](sc: SCollection[A]) {
+
+  def queryRaw(q: String, udfs: Udf*): SCollection[Row] = query(Query(q, udfs.toList))
```

Comment:
I'm not a fan of the `queryRaw*` methods. IMHO they make the API more confusing with the only benefit of saving a few keystrokes.
Using the `Query` object explicitly on the other hand makes it obvious that an instance of `Query` can be shared, for example in a library.

Reply:
I don't have a strong opinion here. @nevillelyh in favor as well in removing these?

Reply:
:+1: on removing. What about a `def query(q: String)` for convenience when there's no UDFs?

Reply:
Or just rename `queryRaw` to `query` and keep the `Udf*`.

Reply:
Yeah convenience was what led me to keep them